eva/optimizer/rules/rules.py

```diff
@@ -413,6 +418,114 @@ def apply(self, before: LogicalFilter, context: OptimizerContext):
         return root_node
 
 
+class CombineSimilarityOrderByAndLimitToFaissIndexScan(Rule):
+    """
+
+    Limit(10)
+        |
+    OrderBy(func)        ->        IndexScan(10)
+        |
+        A
+    """
+
+    def __init__(self):
+        pattern = Pattern(OperatorType.LOGICALLIMIT)
+        orderby_pattern = Pattern(OperatorType.LOGICALORDERBY)
+        orderby_pattern.append_child(Pattern(OperatorType.DUMMY))
+        pattern.append_child(orderby_pattern)
+        super().__init__(
+            RuleType.COMBINE_SIMILARITY_ORDERBY_AND_LIMIT_TO_FAISS_INDEX_SCAN, pattern
+        )
+
+        # Entries populate after rule egibility validation.
+        self._index_catalog_entry = None
```

Comment:
We should not do this as a rule is supposed to be stateless. You can move the check code to the `apply.` We can easily support cases where the rule doesn't return any plan. I made the required changes in my other PR. I will merge it so that you can leverage it.

Reply:
I see. I can move some code to `apply`. Then, when are we supposed to use check?

Reply:
To cater invalidations based on simple logic. It isn't that important in our setting since we don't have a lot of rules.